File: Dopamine.Controls/Tile.vb
Imports System.ComponentModel

Public Class Tile
    Inherits Label

#Region "Variables"
    Private mTile As Border
#End Region

#Region "Properties"
    <EditorBrowsable(EditorBrowsableState.Never)> _
    Public Property IconSize As Double
        Get
            Return CDbl(GetValue(IconSizeProperty))
        End Get

        Set(ByVal value As Double)
            SetValue(IconSizeProperty, value)
        End Set
    End Property

    Public Property IconSizePercent As Double
        Get
            Return CDbl(GetValue(IconSizePercentProperty))
        End Get

        Set(ByVal value As Double)
            SetValue(IconSizePercentProperty, value)
        End Set
    End Property
#End Region

#Region "Dependency properties"
    Public Shared ReadOnly IconSizeProperty As DependencyProperty = DependencyProperty.Register("IconSize", GetType(Double), GetType(Tile), New PropertyMetadata(Nothing))
    Public Shared ReadOnly IconSizePercentProperty As DependencyProperty = DependencyProperty.Register("IconSizePercent", GetType(Double), GetType(Tile), New PropertyMetadata(Nothing))
#End Region

#Region "Construction"
    Shared Sub New()
        'This OverrideMetadata call tells the system that this element wants to provide a style that is different than its base class.
        'This style is defined in themes\generic.xaml
        DefaultStyleKeyProperty.OverrideMetadata(GetType(Tile), New FrameworkPropertyMetadata(GetType(Tile)))
    End Sub
#End Region

#Region "Functions"
    Public Overrides Sub OnApplyTemplate()
        MyBase.OnApplyTemplate()

        Me.mTile = DirectCast(GetTemplateChild("PART_Tile"), Border)

        If Me.mTile IsNot Nothing Then
            Me.SetIconSize(Me.mTile)

            AddHandler Me.mTile.SizeChanged, AddressOf SizeChangedHandler
        End If
    End Sub

    Private Sub SizeChangedHandler(sender As Object, e As SizeChangedEventArgs)
        Me.SetIconSize(CType(sender, Border))
    End Sub

    Private Sub SetIconSize(iTile As Border)
        Try
            ' For some reason, "ActualHeight" is always 0 when arriving here, so we use "Height"
            Me.IconSize = iTile.Height * (Me.IconSizePercent / 100)
        Catch ex As Exception
            Me.IconSize = 0
        End Try
    End Sub
#End Region
End Class

File: Dopamine/Controls/Tile.cs
using System;
using System.ComponentModel;
using System.Windows;
using System.Windows.Controls;

namespace Dopamine.Controls
{
    public class Tile : Label
    {
        private Border mTile;

        [EditorBrowsable(EditorBrowsableState.Never)]
        public double IconSize
        {
            get { return Convert.ToDouble(GetValue(IconSizeProperty)); }

            set { SetValue(IconSizeProperty, value); }
        }

        public double IconSizePercent
        {
            get { return Convert.ToDouble(GetValue(IconSizePercentProperty)); }

            set { SetValue(IconSizePercentProperty, value); }
        }

        public static readonly DependencyProperty IconSizeProperty = DependencyProperty.Register("IconSize", typeof(double), typeof(Tile), new PropertyMetadata(null));
        public static readonly DependencyProperty IconSizePercentProperty = DependencyProperty.Register("IconSizePercent", typeof(double), typeof(Tile), new PropertyMetadata(null));

        static Tile()
        {
            DefaultStyleKeyProperty.OverrideMetadata(typeof(Tile), new FrameworkPropertyMetadata(typeof(Tile)));
        }

        public override void OnApplyTemplate()
        {
            base.OnApplyTemplate();

            this.mTile = (Border)GetTemplateChild("PART_Tile");

            if (this.mTile != null)
            {
                this.SetIconSize(this.mTile);

                this.mTile.SizeChanged += SizeChangedHandler;
            }
        }

        private void SizeChangedHandler(object sender, SizeChangedEventArgs e)
        {
            this.SetIconSize((Border)sender);
        }

        private void SetIconSize(Border iTile)
        {
            try
            {
                // For some reason, "ActualHeight" is always 0 when arriving here, so we use "Height"
                this.IconSize = iTile.Height * (this.IconSizePercent / 100);
            }
            catch (Exception)
            {
                this.IconSize = 0;
            }
        }
    }
}
